aquadoggo/src/rpc/methods/publish_entry.rs

```diff
@@ -48,22 +45,10 @@ pub async fn publish_entry(
     let author = params.entry_encoded.author();
 
     // Determine expected log id for new entry: a `CREATE` entry is always stored in the next free
-    // user log. An `UPDATE` or `DELETE` message is always stored in the same log that its original
+    // log. An `UPDATE` or `DELETE` message is always stored in the same log that its original
     // `CREATE` message was stored in.
-    let document_log_id = match message.is_create() {
-        true => {
-            // A document is identified by the hash of its `CREATE` message
-            let document_hash = &params.entry_encoded.hash();
-            Log::find_document_log_id(&pool, &author, document_hash).await?
-        }
-        false => {
-            // An instance is identified by the hash of its previous operation
```

Comment:
Maybe I'm misunderstanding this here but I stumbled upon this line and the regarding `get_log_id_by_instance` method. Is this the right definition of *instance*? I thought an Instance is a materialization of a Document and it is (still) identified by the hash of its `CREATE` message.

Also the implementation of `get_log_id_by_instance` returns a log id for any event hash which might not necessarily be one from an Instance (that event hash might just be holding an UPDATE Operation). I thought the name `get_log_id_by_entry` is more accurate here then.

Again, could be that I missed the crucial part here, let me know what you think!

Reply:
Interesting! I think we should have discussed the definitions [here](https://github.com/p2panda/aquadoggo/issues/43) at the top together to talk about stuff like this.

> 1

An instance is a _specific_ materialisation of a document. A document can have many instances. That's why we can't identify all of those instances by the document id (hash of the create message).

> 2

I thought we can first implement everything (also materialisation, queries, etc) for single-writer because that's the base scenario rather than build everything at once. Then we have a full working stack and can "upgrade" it for the multi-writer scenario.

> 3

I think I confused myself here :D I followed the definition [here](https://github.com/p2panda/aquadoggo/issues/43) and I think reading _An instance is identified by the hash of the last operation applied._ my head went _look for the last->previous hash_. Going by the current operation's entry hash makes more sense.

> 4

Every operation creates a new instance. As all operations are stored in the same log in the single-writer scenario we can use this shortcut

Reply:
Thank you! I'll dive into it!

Would it be possible for you to sketch a client / node communication flow (similar to what we did once) for publishing new entries? Think it can help me to think about all of this closer to implementation.

Reply:
> my head went look for the last->previous hash. Going by the current operation's entry hash makes more sense.

`id` is giving you the document hash (former "instance id" this is where the name comes from), not the one of the current operation entry (this would be `hash`), maybe this is the confusion here?

If you would go by the current operation entry hash in that code you would get a new log id for every entry, regardless if they belong to the same document or not.

Reply:
> 1
I had actually been imagining that the `Instance` would still be identified by the hash of the CREATE entry. I need to think more deeply though... Definitely worth discussing this together. Apart from that I had the same understanding of `Instance` @cafca

Reply:
> I think we should have discussed the definitions here at the top together to talk about stuff like this.

All the definitions make sense (except of maybe that an `Operation` is not encoded as an `Entry`, Operations are encoded as CBOR and always come *together* with an Entry which provides its integrity and authenticity, the hash of the attached Entry is also used to identify that Operation). Also I understood that `Instance` is still using the same hash as an identifier as the Document.

Instance doesn't seem to be a good word (I like "View" or "DocumentView") and also in general I find that concept important in theory but almost nothing a client will think about, especially when the hash is the same, its a detail I will not care about when querying the database via GraphQL (you could see it also as "I'm querying a Document in the eventually consistent state my Node has").

It gets harder since Instances are something which can only exist *after* materialization. I think this is also why I got confused seeing it here, since we don't have materialization implemented yet.

> I think I confused myself here :D I followed the definition here and I think reading An instance is identified by the hash of the last operation applied. my head went look for the last->previous hash. Going by the current operation's entry hash makes more sense.

More confusion analysis! We talked about "Log per Instance" for a long while *before* we changed the definition of an Instance. It should be called "Log per Document"! I could see that you changed the tests in a way where every newly created operation will lead to a new log id even though they belong to the same document, the code always uses a new hash and therefore a new log id.

Here maybe a small graphic on how I believe it is correct (let me know if that is false):

```
Document x:
log 1: [x] <- [y] <- [z] <- [l]              (key A)
log 2:          \- [i] <- [j] <- [k]         (key B)
```

This is how it is currently implemented after the "Instance per Log" changes:

```
log 1: [x]                                     (key A)
log 2:   \- [y]                                (key A)
log 3:        \- [z]                           (key A)
log 3:            \- [l]                       (key A)
```

Another thing which might have been confusing is the `id` in the `EncodedEntry` struct used in the code. It does not return the entry hash (this would be `hash`), but the document id! I guess we should call it `documentHash` or similar in the future?

> I thought we can first implement everything (also materialisation, queries, etc) for single-writer because that's the base scenario rather than build everything at once. Then we have a full working stack and can "upgrade" it for the multi-writer scenario.

It could make sense to already do *some* things in multi-writer, I actually think especially this flow should already cover it since materialization is relatively hard to think of without multi-writer (its fully designed and already supports multi-writer in the implementation). Multi-writer gets more complicated with Permissions etc., but just the materialization itself is already there :sparkles:

Reply:
(nice graphics :heart: )

Reply:
I really like this concept we have of being able to derive many `Instances` (I don't mind the naming, could be any of your suggestions too @adzialocha) from a single document. However I was thinking of it more as a conceptual/internal concept, and that clients would be making requests to `Documents` and getting back the most up-to-date `Instance` (materialisation). So actually they wouldn't know so much about the distinction and only one materialised result exists at a time.